### PR TITLE
MediaType as type of media content

### PR DESCRIPTION
**Pull request by kbats183**

### CODE REVIEW

File: backend/src/main/kotlin/org/icpclive/controllers/SingleWidgetController.kt [hunk 1]
```diff
@@ -31,8 +31,10 @@ open class SingleWidgetController<SettingsType : ObjectSettings, DataType : Type
         widgetConstructor(settings)
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/controllers/SingleWidgetController.kt:31]: Probably, we should also use constructWidget here

File: backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt [hunk 2]
```diff
@@ -1,26 +1,51 @@
...
-    SingleWidgetController<TeamViewSettings, TeamViewWidget>(TeamViewSettings(), manager, ::TeamViewWidget) {
-    override suspend fun createWidgetAndShow(settings: TeamViewSettings) {
+    SingleWidgetController<ExternalTeamViewSettings, TeamViewWidget>(
+        ExternalTeamViewSettings(),
+        manager,
+        { TeamViewWidget(OverlayTeamViewSettings()) }) {
+    override suspend fun constructWidget(settings: ExternalTeamViewSettings): TeamViewWidget {
+        val teamInfo = DataBus.contestInfoFlow.await().first().teams.find { it.id == settings.teamId }
+        val content = mutableListOf<MediaType>()
+        settings.mediaTypes.forEach {
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt:18]: Isn't it content = settings.MapNotNull?

File: backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt [hunk 3]
```diff
@@ -1,26 +1,51 @@
...
+            val url = teamInfo?.medias?.get(it) ?: return@forEach
+            content += when (it) {
+                TeamMediaType.CAMERA -> MediaType.WebRTCConnection(url)
+                TeamMediaType.SCREEN -> MediaType.WebRTCConnection(url)
+                TeamMediaType.REACTION_VIDEO -> MediaType.Video(url)
+                TeamMediaType.RECORD -> MediaType.Video(url)
+                TeamMediaType.PHOTO -> MediaType.Photo(url)
+                else -> null
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt:26]: You can just do else -> return or else -> continue if replace with just a loop.

File: backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt (hunk 2 above)
Comment (kunyavskiy) [reply backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt:18]: Oh, that's not the only source. So mabye b uildList?

File: backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt [hunk 4]
```diff
@@ -1,26 +1,51 @@
...
 import org.icpclive.data.Manager
 import kotlin.time.Duration.Companion.seconds
 
 class TeamViewController(val manager: Manager<TeamViewWidget>, val position: TeamViewPosition) :
-    SingleWidgetController<TeamViewSettings, TeamViewWidget>(TeamViewSettings(), manager, ::TeamViewWidget) {
-    override suspend fun createWidgetAndShow(settings: TeamViewSettings) {
+    SingleWidgetController<ExternalTeamViewSettings, TeamViewWidget>(
+        ExternalTeamViewSettings(),
+        manager,
+        { TeamViewWidget(OverlayTeamViewSettings()) }) {
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/controllers/TeamViewController.kt:14]: I think this lambda shouldn't be used, as we override construct. So, maybe we should throw an exception here? 

To be honest, I don't like having both this lambda and override, but I have no good idea what to do with it right now.

File: backend/src/main/kotlin/org/icpclive/api/RunInfo.kt [hunk 5]
```diff
@@ -17,6 +17,6 @@ data class RunInfo constructor(
     @Serializable(with = DurationInMillisecondsSerializer::class)
     val time: Duration,
     val isFirstSolvedRun: Boolean = false,
-    val featuredRunMedia: MediaType? = null,
+    val featuredRunMedia: TeamMediaType? = null,
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/api/RunInfo.kt:20]: But where front would convert this media to someting meaningful?

File: overlay/src/components/organisms/widgets/TeamView.js [hunk 6]
```diff
@@ -281,16 +290,47 @@ const TeamVideo = ({ teamId, type, setIsLoaded }) => {
...
+                        muted/>
+                </TeamVideoAnimationWrapper>;
+            }
+            if (c.type === "WebRTCConnection") {
+                return <TeamVideoAnimationWrapper>
+                    <TeamWebRTCVideoWrapper url={c.url} setIsLoaded={setIsLoaded}/>
+                </TeamVideoAnimationWrapper>;
+            }
+            useEffect(() => {
+                setLoadedComponents(m => m | (1 << index));
```
Comment (irdkwmnsb) [inline overlay/src/components/organisms/widgets/TeamView.js:330]: why is this?

File: overlay/src/components/organisms/widgets/TeamView.js [hunk 7]
```diff
@@ -281,16 +290,47 @@ const TeamVideo = ({ teamId, type, setIsLoaded }) => {
...
+                return <TeamVideoAnimationWrapper>
+                    <TeamImageWrapper src={c.url} onLoad={() => setIsLoaded(true)}/>
+                </TeamVideoAnimationWrapper>;
+            }
+            if (c.type === "Video") {
+                return <TeamVideoAnimationWrapper>
+                    <TeamVideoWrapper
+                        src={c.url}
+                        onCanPlay={() => setIsLoaded(true)}
+                        onError={() => setIsLoaded(false) || dispatch(pushLog("ERROR on loading video in TeamView widget"))}
```
Comment (irdkwmnsb) [inline overlay/src/components/organisms/widgets/TeamView.js:319]: setIsLoaded is a function with no arguments.
the argument is ignored, so this won't change what's displayed

Comment (irdkwmnsb) [inline overlay/src/components/organisms/widgets/TeamView.js:319]: Also you shouldn't chain function calls like that.
By semantics setIsLoaded might return true on set success, so no log message will be writter

**Pull request merged**